File: src/pkg/orchestrator/jws.py
# Copyright (c) 2026 NyxeraLabs
# Author: José María Micoli
# Licensed under BSL 1.1
# Change Date: 2033-02-22 -> Apache-2.0
#
# You may:
# Study
# Modify
# Use for internal security testing
#
# You may NOT:
# Offer as a commercial service
# Sell derived competing products

"""Compact JWS payload generation for orchestrator manifests."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any

from pkg.orchestrator.signing import ManifestSigner


class JWSPayloadError(ValueError):
    """Raised when JWS generation cannot complete safely."""


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _json_compact(data: dict[str, Any]) -> bytes:
    return json.dumps(
        data,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=True,
    ).encode("utf-8")


def _decode_signer_signature(signature: str) -> bytes:
    if not signature:
        raise JWSPayloadError("signer returned an empty signature")

    encoded = signature
    if signature.startswith("vault:v"):
        parts = signature.split(":", maxsplit=2)
        if len(parts) != 3:
            raise JWSPayloadError("Vault signature format is invalid")
        encoded = parts[2]

    padding = "=" * (-len(encoded) % 4)
    decoders = (
        lambda value: base64.b64decode(value, validate=True),
        lambda value: base64.b64decode(value, altchars=b"-_", validate=True),
    )
    for decoder in decoders:
        try:
            decoded = decoder(encoded + padding)
        except (ValueError, TypeError):
            continue
        if decoded:
            return decoded

    raise JWSPayloadError("unable to decode signer signature")


@dataclass(slots=True, frozen=True)
class JWSConfig:
    """Config for compact JWS generation."""

    algorithm: str = "ES256"
    key_id: str | None = None
    include_typ: bool = True
    typ: str = "JWT"


class CompactJWSGenerator:
    """Generate compact JWS strings for orchestrator execution payloads."""

    def __init__(self, signer: ManifestSigner, config: JWSConfig | None = None) -> None:
        self._signer = signer
        self._config = config or JWSConfig()

    def generate(self, payload: dict[str, Any]) -> str:
        """Build, sign, and return compact JWS: header.payload.signature."""
        if not payload:
            raise JWSPayloadError("payload must not be empty")

        header: dict[str, Any] = {"alg": self._config.algorithm}
        if self._config.include_typ:
            header["typ"] = self._config.typ
        if self._config.key_id:
            header["kid"] = self._config.key_id

        header_segment = _b64url_encode(_json_compact(header))
        payload_segment = _b64url_encode(_json_compact(payload))
        signing_input = f"{header_segment}.{payload_segment}"

        signer_value = self._signer.sign_payload(signing_input.encode("ascii"))
        signature_segment = _b64url_encode(_decode_signer_signature(signer_value))
        return f"{signing_input}.{signature_segment}"

File: src/runner-go/runner/jws.go
// Copyright (c) 2026 NyxeraLabs
// Author: Jose Maria Micoli
// Licensed under BSL 1.1
// Change Date: 2033-02-22 -> Apache-2.0
//
// You may:
// Study
// Modify
// Use for internal security testing
//
// You may NOT:
// Offer as a commercial service
// Sell derived competing products

package runner

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"strings"
)

var (
	ErrInvalidJWSFormat    = errors.New("invalid compact jws format")
	ErrUnsupportedJWSAlg   = errors.New("unsupported jws algorithm")
	ErrMissingVerifyKey    = errors.New("missing Ed25519 verify key")
	ErrInvalidVerifyKey    = errors.New("invalid Ed25519 verify key")
	ErrSignatureValidation = errors.New("jws signature validation failed")
)

func parseSegment(seg string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(seg)
}

func VerifyEdDSAJWS(compact, publicKeyPEM string) (map[string]any, error) {
	parts := strings.Split(compact, ".")
	if len(parts) != 3 {
		return nil, ErrInvalidJWSFormat
	}
	headerRaw, err := parseSegment(parts[0])
	if err != nil {
		return nil, fmt.Errorf("decode header: %w", err)
	}
	payloadRaw, err := parseSegment(parts[1])
	if err != nil {
		return nil, fmt.Errorf("decode payload: %w", err)
	}
	signatureRaw, err := parseSegment(parts[2])
	if err != nil {
		return nil, fmt.Errorf("decode signature: %w", err)
	}

	var header map[string]any
	if err := json.Unmarshal(headerRaw, &header); err != nil {
		return nil, fmt.Errorf("parse header: %w", err)
	}
	alg, _ := header["alg"].(string)
	if alg != "EdDSA" {
		return nil, ErrUnsupportedJWSAlg
	}
	if publicKeyPEM == "" {
		return nil, ErrMissingVerifyKey
	}

	signingInput := []byte(parts[0] + "." + parts[1])
	block, _ := pem.Decode([]byte(publicKeyPEM))
	if block == nil {
		return nil, ErrInvalidVerifyKey
	}
	parsed, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, ErrInvalidVerifyKey
	}
	publicKey, ok := parsed.(ed25519.PublicKey)
	if !ok {
		return nil, ErrInvalidVerifyKey
	}
	if !ed25519.Verify(publicKey, signingInput, signatureRaw) {
		return nil, ErrSignatureValidation
	}

	var payload map[string]any
	if err := json.Unmarshal(payloadRaw, &payload); err != nil {
		return nil, fmt.Errorf("parse payload: %w", err)
	}
	return payload, nil
}
